doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,824 @@
+  second [label="x.2 = b"]
+  third [label="return &Phi;(x.1,x.2)"]
+  first -> second [label="true"]
+  first -> third [label="false"]
+  second -> third
+}
+\enddot
+
+In CBMC, we provide a precise implemenation of &Phi;, using explicitly tracked
+information about which branches were taken by the program.
+There are also some differences in how loops are
+handled (finite unrolling in CBMC, versus a &Phi;-based approach in
+compilers); the CBMC approach will be discussed in a later chapter.
+
+For the time being, let us come back to `factorial`. We can now give
+an SSA using &Phi; functions:
+
+\dot "Control flow graph in SSA for factorial"
+digraph cfg {
+  node [shape=rectangle]
+  bb1 [label="unsigned long fac.1.1 = 1\nunsigned i.1.1 = 1",peripheries=2]
+  bb2 [label="i.1.2 = &Phi;(i.1.1, i.1.3)\nfac.1.2 = &Phi;(i.1.1, i.1.3)\ni.1.2 <= n.1.1"]
+  bb3 [label="fac.1.3 = fac.1.2 * i.1.2\ni.1.3 = i.1.2 + 1"]
+  
+  bb4 [label="Return fac.1.2",style=filled,fillcolor=gray80]
+  {bb1, bb3} -> bb2
+  bb2 -> bb3 [label="true"]
+  bb2 -> bb4 [label="false"]
+}
+\enddot
+
+The details of SSA construction, plus some discussion of how it is
+used in compilers, can be found in the
+[original paper](http://www.cs.utexas.edu/~pingali/CS380C/2010/papers/ssaCytron.pdf).
+
+The SSA is an extremely helpful representation when one
+wishes to perform model checking on the program (see next section),
+since it is much easier to extract the logic formulas used in this
+technique from an SSA compared to a CFG (or, even worse, an AST). That
+being said, CBMC takes a different route, opting to convert to
+intermediate representation known as GOTO programs instead.
+
+\subsection Trace_section Trace
+
+TODO: I am not quite sure what is supposed to go here.
+
+From GitHub: "I think this was intended to describe how a (counterexample) trace is represented."
 
 \section analysis_techniques_section Analysis techniques
 
 \subsection BMC_section Bounded model checking
 
 To be documented (can copy from the CBMC manual).
 
+TODO: There seems to be no sufficiently comprehensive section in the
+CBMC manual. Find some other sources to collect information from.
+
+- Idea of model checking: Given some property, check if we can reach
+  it; if yes, under what conditions?
+- Bounded model-checking: Cut off search at some point; either prove
+  that property must occur before, or accept underapproximation.
+- How does it work? Encoding of state as propositions/first-order
+  propositions, step function and composition of step functions.
+- Solve question: Is there a model for this formula?
+
 \subsection SAT_section SAT and SMT
 
-To be documented.
+- Goal: Solve model-finding problem.
+- Answers: Here is a model, no model possible
+- Complexity issues
+- SAT and SMT as extremely advanced black boxes that solve
+  a good part of problems efficiently. Point out MiniSat,
+  Z3 as good examples that we use.
+- Pointer to literature: CDCL, Lazy/Eager instantiation
+
```

Comment:
I would keep this section very simple. In addition to the text below I would *only* add the following:

* Give an example of propositional formula
* Give the corresponding DIMACS format that solvers understand. Explain how variables and clauses are encoded.
* Show a shell line how to run minisat on that file, perhaps show the model (it was SAT)